### ISSUE TITLE
about reproduction

### ISSUE THREAD

**Issue by konioy** [ultralytics/yolov5#8633]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Same data, same code, I trained it twice. The loss curve is very close, but the effect on the test data is very different. Why is this？

**Comment by glenn-jocher** [comment]
@konioy current master with torch>=1.12.0 is fully reproducible:
<img width="1252" alt="Screen Shot 2022-07-19 at 6 21 53 PM" src="https://user-images.githubusercontent.com/26833433/179800461-ff6a6b71-0922-4279-80ac-3384219aac57.png">

**Comment by konioy** [comment]
Is it a problem with the pytorch version? Is the 1.9 version of pytorch unreproducible?

**Comment by konioy** [comment]
I looked at (https://github.com/ultralytics/yolov5/pull/8213) and (https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html?highlight=torch%20use_deterministic_algorithms#torch.use_deterministic_algorithms), but don't understand how torch.nn.Upsample is reproducible?

**Comment by glenn-jocher** [comment]
@konioy torch>=1.12 should be fully reproducible using single GPU. Multi-GPU is not yet reproducible and we don't have a clear reason why.

**Comment by konioy** [comment]
I am using pytorch1.9, which I know cannot be reproduced. It may be that the choice of convolution operator is different (because torch.cudnn, benchmark=true).
But from the loss point of view of two repeated training (same code, same data), the difference in loss is very small.
train1 on the left, train2 on the right
![image](https://user-images.githubusercontent.com/24238694/182286885-402d4354-0921-49e2-b48c-ea9af287d271.png)

 But there is a big difference in the validation set. Do you have any ideas or suggestions?
train1:
![resulte_baseline](https://user-images.githubusercontent.com/24238694/182286971-56fc6f15-baa0-42b7-a318-a14793cc9cef.png)
train2
![results](https://user-images.githubusercontent.com/24238694/182286985-2b02581d-eb34-4bdc-909a-5529d7122112.png)

![image](https://user-images.githubusercontent.com/24238694/182287277-155d2d2a-4a6b-4b72-879e-b3da03181195.png)

**Comment by glenn-jocher** [comment]
@konioy zero val loss typically indicates your validation set has no labels

EDIT: if you used --no-val then the above is normal

**Comment by konioy** [comment]
yes, i used --no-val.
I think two repetitions of training, the effect on the validation set should be similar. But my experiments show that repeating the training twice, there is indeed a big gap, as shown below, do you have any suggestions?
**map 0.5:**
train1:0.8162
train2:0.9167
**precision:**
train1:0.8159
train2:0.8985
**recall:**
train1:0.7381
train2:0.8517

**Comment by glenn-jocher** [comment]
@konioy use torch>=12.0 for reproducible Single-GPU CUDA trainings runs

**Comment by konioy** [comment]
Thanks, I know.
Do you have any suggestions for my situation?

**Comment by konioy** [comment]
Thanks, I know.
Do you have any suggestions for my situation?

**Comment by konioy** [comment]
Have you ever encountered this situation?

**Comment by konioy** [comment]
have you encountered a similar situation？

**Comment by glenn-jocher** [comment]
@konioy your results are expected. torch<1.12 will not produce reproducible results.

**Comment by konioy** [comment]
However, the results fluctuated greatly.  this is not expexted.

**Comment by glenn-jocher** [comment]
@konioy Apologies for any confusion. It's possible that the fluctuation in results could be due to the non-reproducibility of training runs with torch<1.12. Upgrading to torch>=1.12 should help minimize these fluctuations.
